src/base/bittorrent/tracker.cpp

```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
     // ip address
     announceReq.socketAddress = m_env.clientAddress;
     announceReq.claimedAddress = queryParams.value(ANNOUNCE_REQUEST_IP);
+    // request port
+    announceReq.requestPort =  m_env.clientPort;
 
     // Enforce using IPv4 if address is indeed IPv4 or if it is an IPv4-mapped IPv6 address
     bool ok = false;
     const qint32 decimalIPv4 = announceReq.socketAddress.toIPv4Address(&ok);
-    if (ok)
+    if (ok) {
         announceReq.socketAddress = QHostAddress(decimalIPv4);
+        LogMsg(tr(R"(Embedded Tracker: Announce request from non- plain-IPv6 address %1.)"
+                R"( Forcing use of equivalent plain IPv4 address %2.)")
+                .arg(m_env.clientAddress.toString(), announceReq.socketAddress.toString())
+            , Log::INFO);
+    }
 
     // 1. info_hash
     const auto infoHashIter = queryParams.find(ANNOUNCE_REQUEST_INFO_HASH);
-    if (infoHashIter == queryParams.end())
-        throw TrackerError("Missing \"info_hash\" parameter");
+    if (infoHashIter == queryParams.end()) {
+        LogMsg(tr(R"(Embedded Tracker: Missing info_hash parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"(Missing "info_hash" parameter)");
+    }
 
     const InfoHash infoHash(infoHashIter->toHex());
-    if (!infoHash.isValid())
-        throw TrackerError("Invalid \"info_hash\" parameter");
+    if (!infoHash.isValid()) {
+        LogMsg(tr(R"(Embedded Tracker: Invalid info_hash parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"(Invalid "info_hash" parameter)");
+    }
 
     announceReq.infoHash = infoHash;
 
     // 2. peer_id
     const auto peerIdIter = queryParams.find(ANNOUNCE_REQUEST_PEER_ID);
-    if (peerIdIter == queryParams.end())
-        throw TrackerError("Missing \"peer_id\" parameter");
+    if (peerIdIter == queryParams.end()) {
+        LogMsg(tr(R"(Embedded Tracker: Missing "peer_id" parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"(Missing "peer_id" parameter")");
+    }
 
-    if (peerIdIter->size() > PEER_ID_SIZE)
-        throw TrackerError("Invalid \"peer_id\" parameter");
+    if (peerIdIter->size() > PEER_ID_SIZE) {
+        LogMsg(tr(R"(Embedded Tracker: "peer_id" too big, truncated)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::WARNING);   
+    }
 
     announceReq.peer.peerId = *peerIdIter;
+    announceReq.peer.peerId.truncate(20);
```

Comment:
do we really have to do this?

Reply:
https://github.com/qbittorrent/qBittorrent/pull/12434/files#r404058554

Reply:
IMO we are wasting CPU cycles on something that has no effect (except your mind of "more leniency")...

Reply:
On a second thought, maybe this does make sense, I mean to adhere to BEP. 

>https://github.com/qbittorrent/qBittorrent/pull/12434/files#r404058554

This is not a constructive response. I see you're even lazy to state your reason for the change.

Reply:
you'll need to use `peerIdIter->left(20)` on previous line.

Reply:
@Chocobo1
> This is not a constructive response. I see you're even lazy to state your reason for the change.

Not lazy, I only linked the response because the response would have been the exact same as for other similar comments... would you rather I had copy-pasted the exact same thing multiple times? These are all instances of apparently not-so-compliant cases where the spec does not explicitly mandate the request to fail, and where we can default to a reasonable lenient behavior.

But just to be extremely clear in this specific case:
The spec says the length is 20, so we truncate to that internally (while still logging if this happens), which is a reasonable lenient alternative to just not processing the request at all if the length is not 20, because the spec does not explicitly say "if `peer_id` is not 20 bytes in length, the request should/must fail". In fact, it does not explicitly state what to do at all when the length is not 20.

Reply:
@Chocobo1

> it is better to use `peerIdIter->left(20)` on previous line.

Fine by me. Out of curiosity, what are the benefits?

Reply:
Fixed, by the way.

Reply:
Isn't it obvious? so it would copy a shorter array instead of a longer one.